Persist last selected database in the native query editor for a user

nemanjaglumac opened the pull request:
## Info
|||
|-|-|
| 📝 | [Product Doc](https://www.notion.so/metabase/Default-to-the-last-selected-database-in-native-queries-3bf8f844a6e8407a998bbf11a04f106b) |
| 🧪 | [Testing Plan](https://github.com/metabase/metabase/issues/39035) |

## What does PR accomplish?
This PR resolves #8780 by introducing a user-level setting for remembering the last used database in the native query.
Closes #39035.

```[tasklist]
### Tasks
- [x] Add new user-level setting to the backend
- [x] Wire native query editor to remember the last used database
- [x] Narrow the scope to just native questions and native models
- [x] Write unit tests
- [x] Write E2E tests
```

## Details
### About the `last-used-database` setting
- It should be user-local only
- It should work for any authenticated user; frontend ensures that a user without sufficient native permissions is not even offered to select and persist the last used database
- The information about the last used database currently should not be included in the audit information

> [!Note]
> The backend currently does have any additional checks to make sure the previously used database is still present. Frontend simply falls back to no database selected and prompts the user to select a new one.

nemanjaglumac committed:
Add new user-specific setting `last-used-database-id`

nemanjaglumac committed:
Update user settings when user selects a database for a native query

nemanjaglumac committed:
Remember the last used database in the native query editor

nemanjaglumac committed:
Fix type error

nemanjaglumac committed:
Set database id when initializing new native question

nemanjaglumac committed:
Persist the database only for "New > Native query"

nemanjaglumac committed:
Persist last used database for native models

nemanjaglumac committed:
Mark `last-used-database-id` setting as potentially `undefined`

nemanjaglumac committed:
Reduce diff

nemanjaglumac committed:
Remove unused prop from QB

ranquild commented on frontend/src/metabase/query_builder/components/NativeQueryEditor/NativeQueryEditor.tsx:
```diff
@@ -909,6 +921,14 @@ const mapDispatchToProps = (dispatch: Dispatch) => ({
     );
     return Questions.HACK_getObjectFromAction(action);
   },
+  onSetDatabaseId: (id: DatabaseId) => {
```
Please remove `shouldPersistLastUsedDatabase`, add a callback to `NativeQueryEditor` like `onSetDatabaseId` and put the action in `View` and not here

nemanjaglumac replied:
Help me understand this better.
What are we trying to achieve by moving the dispatch up to the `View`?

The reason I chose this approach was because I wanted a fine-grained control of when exactly do we persist the database. NativeQueryEditor can be rendered by many different components.

How are we going to control when does native query editor persists the database ID, and when does it ignore it?

nemanjaglumac replied:
To answer my own question - this should do it.
https://github.com/metabase/metabase/pull/38986/commits/4dd6aad70815b5780baf5a675b2f62ed98146ebc

I'd still like to learn why is one approach considered better than the other?

ranquild reviewed (commented):
@nemanjaglumac we would also need to handle database ids after the database is removed. I think we should do in clojure in `last-used-database-id` getter and return `null` if the corresponding DB cannot be found (could be done in another PR)

nemanjaglumac committed:
Lift dispatch logic up to the `View` component

nemanjaglumac committed:
Add initial E2E tests

nemanjaglumac committed:
Fix type errors

nemanjaglumac committed:
Make this setting work for all authenticated users

nemanjaglumac committed:
Add initial backend unit test

nemanjaglumac committed:
Fix E2E test and make it more explicit

nemanjaglumac committed:
Add mongo-related E2E test

nemanjaglumac committed:
Fix typo

nemanjaglumac committed:
Add permission-related E2E tests

- Adds reproduction for #39053

nemanjaglumac committed:
Make sure no redundant requests are made

This commit ensures no redundant requests are made when selecting
the same, i.e. the previously selected, database.

nemanjaglumac committed:
Add E2E tests related to model actions

nemanjaglumac committed:
Clean up

nemanjaglumac committed:
Extract persist database logic into an action

kamilmielnik commented on e2e/test/scenarios/native/native-database-source.cy.spec.js:
```diff
@@ -0,0 +1,247 @@
+  POPOVER_ELEMENT,
+} from "e2e/support/helpers";
+
+const PG_DB_ID = 2;
```
Isn't this `WRITABLE_DB_ID` ?

nemanjaglumac replied:
In theory yes.
But it can be misleading. We're using `WRITABLE_DB_ID` only in the context of actual writable scenarios, and we're not writing anything to this database.

I'm working on a better solution for all QA DB references, btw.

kamilmielnik commented on e2e/test/scenarios/native/native-database-source.cy.spec.js:
```diff
@@ -0,0 +1,247 @@
+});
+
+function startNativeQuestion() {
+  cy.visit("/");
```
This is expensive. Can we use SPA navigation instead?

```suggestion
  cy.findByTestId("main-logo-link").click();
```

----

The same comment applies to other `cy.visit` calls in this test suite.

nemanjaglumac replied:
It is expensive, but it was the only reliable approach that gave me confidence this thing works.
I can explore some other options.

kamilmielnik commented on e2e/test/scenarios/native/native-database-source.cy.spec.js:
```diff
@@ -0,0 +1,247 @@
+        cy.request("DELETE", `/api/database/${databaseId}`);
+      });
+
+      startNativeQuestion();
+      assertNoDatabaseSelected();
```
I wonder if this test is going to pass if `startNativeQuestion` does not call `cy.visit`.
It should not, and we should ensure that the setting value is not stale (i.e. is refetched) when visiting native query editor.

ranquild replied:
+1

nemanjaglumac replied:
Unfortunately, our caching logic creates all sorts of issues not just here, but in a lot of other places. Unless we do a hard refresh (which visiting a home page achieves in Cypress), most of this will not work.

This also applies to [the other comment](https://github.com/metabase/metabase/pull/38986#discussion_r1503801570) and [my answer](https://github.com/metabase/metabase/pull/38986#discussion_r1503871038) there.

kamilmielnik commented on e2e/test/scenarios/native/native-database-source.cy.spec.js:
```diff
@@ -0,0 +1,247 @@
+    });
+
+    describe("permissions", () => {
+      it("users with 'No self-service' data permissions should be able to choose only the databases they can query against", () => {
```
I think this scenario (similar to when DB is removed) is also worth testing:

1. User has access to DB X
2. User creates a native query based on DB X
3. User is revoked permissions to DB X
4. User opens native query editor

Assertion: DB X is not listed and not selected

nemanjaglumac replied:
Added a new test for this.
As with the other scenarios, unless there is a hard refresh user will still see the old database.

Spoke to AlexP about this and we concluded that this is a global problem we have - definitely not tied to just this PR.

kamilmielnik commented on src/metabase/models/user.clj:
```diff
@@ -439,6 +439,12 @@
   :user-local :only
   :type :string)
 
+(defsetting last-used-database-id
+  (deferred-tru "The last database a user has used for a native query.")
```
I suggest to be more specific with the name.

```suggestion
(defsetting last-used-native-query-database-id
  (deferred-tru "The last database a user has used for a native query.")
```

kamilmielnik reviewed (commented):
Beautifully written test suite <3

nemanjaglumac commented on frontend/src/metabase/query_builder/components/DataSelector/TriggerComponents.tsx:
```diff
@@ -80,7 +80,12 @@ export function FieldTrigger({
-    <span className="text-wrap text-grey no-decoration">{database.name}</span>
+    <span
+      className="text-wrap text-grey no-decoration"
+      data-testid="selected-database"
```
Yes, it is the same because they can never be rendered at the same time. It's either-or type of situation, but the meaning of that element is the same - it shows the selected database.

1. when it's the only option and thus rendered by default
2. when it's the chosen option out of many

nemanjaglumac commented on frontend/src/metabase/query_builder/components/view/View.jsx:
```diff
@@ -397,4 +403,11 @@ class View extends Component {
+export default _.compose(
+  ExplicitSize({ refreshMode: "debounceLeading" }),
+  connect(null, mapDispatchToProps),
+)(View);
```
It's funny cause [I used that exact same logic originally](https://github.com/metabase/metabase/pull/38986/commits/bffccafc66dd8b30707ce5f4724512e9d1e35bd8) (defined the action in the `NativeQueryEditor`) but @ranquild [asked me](https://github.com/metabase/metabase/pull/38986#discussion_r1499174779) to lift it to the `View`.

ranquild replied:
We should `connect` early. The current QueryBuilder/View/ViewHeader structure is broken. This is closer to the correct way of doing it

nemanjaglumac commented on frontend/src/metabase-types/api/settings.ts:
```diff
@@ -285,6 +285,7 @@ export interface Settings {
   "user-visibility": string | null;
   "last-acknowledged-version": string | null;
   "show-static-embed-terms": boolean | null;
+  "last-used-database-id"?: number | null;
```
Great question @kamilmielnik !
This is something that I didn't know about, and it's not documented anywhere.

@ranquild explained it, and I plan to document it AND to update types to reflect this. But I thought it falls outside the scope of this PR.

tl;dr
`null` or `value` - user has access and it reads the actual value
`undefined` - user does not have permissions to fetch this

ranquild commented on frontend/src/metabase-types/api/mocks/settings.ts:
```diff
@@ -227,5 +227,6 @@ export const createMockSettings = (
   "uploads-schema-name": null,
   "user-visibility": null,
   "last-acknowledged-version": "v1",
+  "last-used-database-id": 1,
```
I wouldn't do that. Technically these presents are not connected

ranquild replied:
E.g. leave 1

ranquild reviewed (commented):
Looks good. I'm worried that the case with deleting the last selected database doesn't work well. There are some comments from @kamilmielnik about that. Could you double check?

oleggromov commented on test/metabase/models/user_test.clj:
```diff
@@ -194,12 +194,12 @@
+      (mt/with-temporary-setting-values [enable-password-login false]
+        (ldap.test/with-ldap-server
+          (invite-user-accept-and-check-inboxes! :invitor default-invitor , :accept-invite? false)
+          (is (seq (mt/regex-email-bodies #"/auth/login"))))))))
```
My 2c about changes like that — I am not sure which linters/prettifiers everyone uses for clojure, but probably different. Let's avoid changes like that because they unnecessarily change the history.

nemanjaglumac replied:
Oopsy.
I didn't even notice that because I have the whitespace turned off in GH UI.

Thanks

nemanjaglumac committed:
Expand the backend unit test to check for user-local only property

nemanjaglumac committed:
Update the setting name as suggested in the code review

nemanjaglumac committed:
Make the setting description a bit more explicit

nemanjaglumac committed:
Add one more permissions-based E2E test

nemanjaglumac committed:
Store postgres database name in a variable

calherries reviewed (commented):
What happens when the last used database is deleted? Do we need a test for that?

Is there a FE benefit of implementing `last-used-native-database-id` as a user-local setting, instead of a new column of the `core_user` table? (fetched with `GET /api/user/current` and updated with `PUT /api/user/:id`)?

From my perpective on the BE, the advantage of adding a new column in the `core_user` table over using a setting is that it can be a foreign key to the database ID. Then we can do things like automatically set the value to NULL if the database is deleted.

calherries reviewed (commented):
What happens when the last used database is deleted? Do we need a test for that?

Is there a FE benefit of implementing `last-used-native-database-id` as a user-local setting, instead of a new column of the `core_user` table? (fetched with `GET /api/user/current` and updated with `PUT /api/user/:id`)?

From my perpective on the BE, the advantage of adding a new column in the `core_user` table over using a setting is that it can be a foreign key to the database ID. Then we can do things like automatically set the value to NULL if the database is deleted.

calherries commented:
Answering the bonus question here:

> Would it be ok to define a custom :getter for this new defsetting value that would first check if the database still exists AND if the current user has necessary permissions to query it?

Yes. We could also do this in `GET /api/user/current`

nemanjaglumac commented:
> What happens when the last used database is deleted? Do we need a test for that?

@calherries 
We already have it [here](https://github.com/metabase/metabase/pull/38986/files#diff-7140ba4e9ce90fa912c83d5042756afd2c4ee3fb0369b0e078683027ca7b2742R59). Frontend will prompt the user to select a new database.

calherries reviewed (approved):
Looks good enough to merge for me like this anyway, it's more of a nit if you're handling this on the backend. Tests are tight too
